pkg/image/importer/credentials_test.go

```diff
@@ -59,3 +58,53 @@ func TestBasicCredentials(t *testing.T) {
 		t.Fatalf("unexpected response: %s %s", u, p)
 	}
 }
+
+func Test_basicCredentialsFromKeyring(t *testing.T) {
+	fn := func(host string, entry credentialprovider.DockerConfigEntry) credentialprovider.DockerKeyring {
+		k := &credentialprovider.BasicDockerKeyring{}
+		k.Add(map[string]credentialprovider.DockerConfigEntry{host: entry})
+		return k
+	}
+	def := credentialprovider.DockerConfigEntry{
+		Username: "local_user",
+		Password: "local_pass",
+	}
+	type args struct {
+		keyring credentialprovider.DockerKeyring
+		target  *url.URL
+	}
+	tests := []struct {
+		name     string
+		args     args
+		user     string
+		password string
+	}{
+		{name: "exact", args: args{keyring: fn("localhost", def), target: &url.URL{Host: "localhost"}}, user: def.Username, password: def.Password},
+		{name: "https scheme", args: args{keyring: fn("localhost", def), target: &url.URL{Scheme: "https", Host: "localhost"}}, user: def.Username, password: def.Password},
+		{name: "canonical https", args: args{keyring: fn("localhost", def), target: &url.URL{Scheme: "https", Host: "localhost:443"}}, user: def.Username, password: def.Password},
+		{name: "only https", args: args{keyring: fn("https://localhost", def), target: &url.URL{Host: "localhost"}}, user: def.Username, password: def.Password},
+		{name: "only https scheme", args: args{keyring: fn("https://localhost", def), target: &url.URL{Scheme: "https", Host: "localhost"}}, user: def.Username, password: def.Password},
+		{name: "mismatched scheme - http", args: args{keyring: fn("http://localhost", def), target: &url.URL{Scheme: "https", Host: "localhost"}}, user: def.Username, password: def.Password},
+
+		// this is not allowed by the credential keyring, possibly because of insufficient testing
```

Comment:
Maybe because `https` targets get trimmed before a lookup: https://github.com/openshift/origin/pull/15945/commits/8ec0656ca97a809e75f9780a4c14448a773a19ef#diff-d60fc09a78ae86f4931185b756b2ba00R168

Reply:
I meant that the upstream credential keyring is incorrectly handling http credentials.

Reply:
The upstream keyring [cannot handle](https://github.com/openshift/origin/blob/8ec0656ca97a809e75f9780a4c14448a773a19ef/vendor/k8s.io/kubernetes/pkg/credentialprovider/keyring.go#L191) targets with scheme. If the `"https"` hadn't been removed from the target before a call to `Lookup()`, no match would had been found as well. The target shouldn't contain the scheme at all.